### PR TITLE
Vinai/create database

### PR DESCRIPTION
**Pull request by VinaiRachakonda**
Add functionality for `CREATE DATABASE`

### CODE REVIEW

File: sql/plan/dbddl.go [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+	}
+
+	return sql.RowsToRowIter(), nil
+}
+
+func (d DropDB) WithChildren(children ...sql.Node) (sql.Node, error) {
+	return NillaryWithChildren(d, children...)
+}
+
+func NewDropDatabase(dbName string, ifExists bool, collate string, charset string) *DropDB {
```
Comment (zachmu) [inline sql/plan/dbddl.go:141]: Same comment about collation and charset

**Review by zachmu — COMMENTED** [review]
Good start but missing tests. We need:

* parse tests
* engine tests (probably best to stick this kind of thing either in script_queries.go or else as their own standalone methods in enginetests.go)

File: sql/catalog.go [hunk 2]
```diff
@@ -120,6 +127,10 @@ func (d Databases) Database(name string) (Database, error) {
 		return nil, ErrDatabaseNotFound.New(name)
 	}
 
+	if len(name) == 0 {
+		return nil, ErrNoDatabaseSelected.New()
```
[the remark this answers is not in the document]
Comment (VinaiRachakonda) [reply sql/catalog.go:131]: So the problem I was having is that mysql should return a database not set error when you drop the database you are currently using. This adds that functionality. Lmk if that seems like a hacky solution to you.

File: sql/plan/dbddl.go [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+package plan
+
+import (
+	"fmt"
+	"github.com/dolthub/go-mysql-server/memory"
+	"github.com/dolthub/go-mysql-server/sql"
+	"github.com/dolthub/vitess/go/vt/sqlparser"
+)
+
+type CreateDB struct {
```
[the remark this answers is not in the document]
Comment (VinaiRachakonda) [reply sql/plan/dbddl.go:24]: Wdym the integrators?

File: sql/catalog.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (VinaiRachakonda) [reply sql/catalog.go:131]: I'll add a test to sql-server in dolt to show this.

File: enginetest/enginetests.go [hunk 4]
```diff
@@ -1736,6 +1736,196 @@ func TestDropColumn(t *testing.T, harness Harness) {
...
+			"DROP DATABASE mydb",
+			[]sql.Row{{sql.OkResult{RowsAffected: 1}}},
+			nil,
+		)
+
+		_, err := e.Catalog.Database("mydb")
+		require.Error(t, err)
+
+		// TODO: Deal with handling this error.
+		//AssertErr(t, e, harness, "SHOW TABLES", sql.ErrNoDatabaseSelected)
```
Comment (VinaiRachakonda) [inline enginetest/enginetests.go:1866]: @zachmu Are you fine with me leaving this as an issue an moving on? Getting this render correctly seems to be a pain to get right and is not an important feature imo.

Comment (zachmu) [reply enginetest/enginetests.go:1866]: Yeah I think that's fine for now

File: enginetest/enginetests.go [hunk 5]
```diff
@@ -1736,6 +1736,196 @@ func TestDropColumn(t *testing.T, harness Harness) {
...
+	})
+
+	t.Run("DROP DATABASE works on newly created tables.", func(t *testing.T) {
```
Comment (zachmu) [inline enginetest/enginetests.go:1869]: newly created databases?

File: enginetest/enginetests.go [hunk 6]
```diff
@@ -1736,6 +1736,196 @@ func TestDropColumn(t *testing.T, harness Harness) {
...
+		TestQuery(t, harness, e,
+			"DROP DATABASE testdb",
+			[]sql.Row{{sql.OkResult{RowsAffected: 1}}},
+			nil,
+		)
+
+		AssertErr(t, e, harness, "USE testdb", sql.ErrDatabaseNotFound)
+	})
+
+	t.Run("DROP SCHEMA works on newly created tables.", func(t *testing.T) {
```
Comment (zachmu) [inline enginetest/enginetests.go:1888]: created database

File: enginetest/enginetests.go [hunk 7]
```diff
@@ -1736,6 +1736,196 @@ func TestDropColumn(t *testing.T, harness Harness) {
...
+		TestQuery(t, harness, e,
+			"DROP SCHEMA testdb",
+			[]sql.Row{{sql.OkResult{RowsAffected: 1}}},
+			nil,
+		)
+
+		AssertErr(t, e, harness, "USE testdb", sql.ErrDatabaseNotFound)
+	})
+
+	t.Run("DROP DATABASE IF EXISTS correctly works.", func(t *testing.T) {
```
Comment (zachmu) [inline enginetest/enginetests.go:1907]: You need to also do this on a database that doesn't exist and assert no err

File: enginetest/enginetests.go [hunk 8]
```diff
@@ -1736,6 +1736,196 @@ func TestDropColumn(t *testing.T, harness Harness) {
...
+		AssertWarning(t, e, harness, "DROP DATABASE IF EXISTS mydb", 1007)
```
Comment (zachmu) [inline enginetest/enginetests.go:1908]: Use consts for expected err codes

File: sql/plan/dbddl.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (zachmu) [reply sql/plan/dbddl.go:25]: integrators are people who provide database implementation, e.g. dolt

File: sql/plan/dbddl.go [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+	return nil
+}
+
+func (d DropDB) RowIter(ctx *sql.Context, row sql.Row) (sql.RowIter, error) {
+	exists := d.Catalog.HasDB(d.dbName)
+	if !exists {
+		if d.IfExists {
+			ctx.Session.Warn(&sql.Warning{
+				Level:   "Note",
+				Code:    1007,
```
Comment (zachmu) [inline sql/plan/dbddl.go:119]: Use the vitess const for this

**Review by zachmu — APPROVED** [review]
Looks great, just a few comments.

**Pull request merged**